### PR TITLE
New Performance/DoubleStartEndWith cop

### PR DESCRIPTION
**Pull request by DNNX**
This is a continuation of https://github.com/bbatsov/rubocop/pull/2535 .

This cop checks for double `#start_with?` or `#end_with?` calls
separated by `||`. In some cases such calls can be replaced
with an single `#start_with?`/`#end_with?` call.

```ruby
  # bad
  str.start_with?("a") || str.start_with?(Some::CONST)
  str.start_with?("a", "b") || str.start_with?("c")
  var1 = ...
  var2 = ...
  str.end_with?(var1) || str.end_with?(var2)

  # good
  str.start_with?("a", Some::CONST)
  str.start_with?("a", "b", "c")
  var1 = ...
  var2 = ...
  str.end_with?(var1, var2)
```

The corner cases are when arguments to start_with? have side effects,
or when the receiver itself has side effects, or when the receiver
is not a string at all, or when someone monkey-patched `String#start_with?`

Benchmark results are not too convincing because of a big variance, but they show that the single-call approach is at least not slower that the two-call approach.

```ruby
prefix1 = "lib/"
prefix2 = "test/"
{
  'First prefix matches' => 'lib/file.rb',
  'Second prefix matches' => 'test/file.rb',
  'None matches' => 'app/models/file.rb'
}.each do |description, str|
  Benchmark.ips do |x|
    puts "==== #{description} ===="
    x.report('two calls') { str.start_with?(prefix1) || str.start_with?(prefix2) }
    x.report('one call')  { str.start_with?(prefix1, prefix2) }
    x.compare!
  end
end
```

```
==== First prefix matches ====
Calculating -------------------------------------
           two calls    31.010k i/100ms
            one call    35.180k i/100ms
-------------------------------------------------
           two calls      2.035M (±14.6%) i/s -      9.644M
            one call      2.039M (±15.9%) i/s -      9.569M

Comparison:
            one call:  2038728.5 i/s
           two calls:  2034701.2 i/s - 1.00x slower

==== Second prefix matches ====
Calculating -------------------------------------
           two calls    31.139k i/100ms
            one call    34.518k i/100ms
-------------------------------------------------
           two calls      1.515M (±20.7%) i/s -      6.944M
            one call      1.709M (±19.5%) i/s -      7.836M

Comparison:
            one call:  1708570.7 i/s
           two calls:  1514975.3 i/s - 1.13x slower

==== None matches ====
Calculating -------------------------------------
           two calls    33.038k i/100ms
            one call    36.789k i/100ms
-------------------------------------------------
           two calls      1.471M (±18.9%) i/s -      6.773M
            one call      1.778M (±19.9%) i/s -      8.130M

Comparison:
            one call:  1778026.2 i/s
           two calls:  1471409.3 i/s - 1.21x slower
```

Wild life occurrences: https://github.com/ManageIQ/manageiq/pull/5558, https://github.com/rails/rails/pull/22374, the aforementioned https://github.com/bbatsov/rubocop/pull/2535.

### CODE REVIEW

**Comment by alexdowad** [comment]
Nice!! I was thinking of doing this myself, good to see that someone else did it first!

File: lib/rubocop/cop/performance/double_start_end_with.rb [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+      #
+      #   @good
+      #   str.start_with?("a", Some::CONST)
+      #   str.start_with?("a", "b", "c")
+      #   var1 = ...
+      #   var2 = ...
+      #   str.end_with?(var1, var2)
+      class DoubleStartEndWith < Cop
+        MSG = 'Use `str.%{method}(x, ..., y, ...)` ' \
+              'instead of `str.%{method}(x, ...) || str.%{method}(y, ...)`.'
```
Comment (alexdowad) [inline lib/rubocop/cop/performance/double_start_end_with.rb]: Rather than showing `x, ...` and `y, ...`, it would be nicer if the message showed the *actual* arguments.

Comment (DNNX) [inline lib/rubocop/cop/performance/double_start_end_with.rb]: Good point. I'll fix it.

File: lib/rubocop/cop/performance/double_start_end_with.rb [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+        def on_or(node)
+          method1, args1, method2, args2 = two_start_end_with_calls(node)
+          if args1 &&
+             method1 == method2 &&
```
Comment (alexdowad) [inline lib/rubocop/cop/performance/double_start_end_with.rb]: You could eliminate this check if you wanted to by making the pattern like:

```
(or (send _recv [{:start_with? :end_with?} _method] $...) (send _recv _method $...))
```

...whether you think that is better or not is up to you.

Comment (DNNX) [inline lib/rubocop/cop/performance/double_start_end_with.rb]: Yea, I was trying to find a way to unify two `{:start_with? :end_with?}` matchers, but looking at the documentation haven't found any. Thanks for the pointer, I'm going to update the pattern and remove `method1==method2` runtime check. This `if` condition is too big to my taste.

File: lib/rubocop/cop/performance/double_start_end_with.rb [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+             args1.all?(&:pure?) &&
+             args2.all?(&:pure?)
```
Comment (alexdowad) [inline lib/rubocop/cop/performance/double_start_end_with.rb]: When I saw this PR, I wondered if you knew about `Node#pure?` or not... glad to see you found it!

Comment (DNNX) [inline lib/rubocop/cop/performance/double_start_end_with.rb]: Well it was actually you who showed me it :)

**Comment by alexdowad** [comment]
One other way this could be improved -- think of things like this:

```ruby
# this will parse as (or (or something str.start_with?('a')) str.start_with?('b'))...
# in other words the 2 nodes you are interested in will not be nested under the same 'or' node
something || str.start_with?('a') || str.start_with?('b')

# another example:
str.start_with?('a') || foo || bar || str.start_with?('b')
```

Evaluation is from left to right, so you have to make sure that there is no node with unknown side effects between the 2 matching calls.

**Comment by DNNX** [comment]
`something || str.start_with?('a') || str.start_with?('b')` - yea, I was thinking about this, but found it kind of complex to implement so I decided to go with more simple implementation which covers most real-life cases (in my experience). I also saw `!str.start_with?(a) && !str.start_with?(b)` in one open-source project. I can also be simplified, but again, I decided not to go this far. Do you think it should be done in this PR?

Another thing which can potentially be done is we can loosen the requirement of the first argument to the first `start_with?` call to be pure. As long as all other arguments are pure, the first one can have side effects  - the semantics will stay the same.

**Comment by alexdowad** [comment]
> Do you think it should be done in this PR?

It's up to you. I am just suggesting ways to make the PR better. If you don't want to go that far in this PR, and the maintainers want to accept it as is, that is fine. It is already an improvement over what we have.

**Comment by DNNX** [comment]
Here is a little summary:

- [x] Implemented more intelligent matcher as per @alexdowad's suggestion
- [x] Improved offense message format. No more abstract `x, ... y, ...`; the message now includes the actual code
- [x] Dropped the requirement for the arguments of the first `start_with?` call to be pure
- [ ] `something || str.start_with?('a') || str.start_with?('b')` - not going to do this in this PR if you guys don't mind
- [ ] `str.start_with?('a') || foo || bar || str.start_with?('b')` - will leave it as well
- [ ] `!str.start_with?(a) && !str.start_with?(b)` - not going to do this either (in this PR at least)

If you guys are ok with this, please let me know. I will squash the commits and add an entry to the change log then.

**Comment by DNNX** [comment]
UPD: This PR is ready to merge now, if everyone is ok with the changes.

**Comment by jonas054** [comment]
:+1: The changes look good.

File: lib/rubocop/cop/performance/double_start_end_with.rb [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+      #   str.start_with?("a", "b", "c")
+      #   var1 = ...
+      #   var2 = ...
+      #   str.end_with?(var1, var2)
+      class DoubleStartEndWith < Cop
+        MSG = 'Use `%{receiver}.%{method}(%{combined_args})` ' \
+              'instead of `%{original_code}`.'
+
+        def on_or(node)
+          receiver, method, args1, args2 = two_start_end_with_calls(node)
```
Comment (bbatsov) [inline lib/rubocop/cop/performance/double_start_end_with.rb]: args1 and args2 are not very descriptive variable names.

File: lib/rubocop/cop/performance/double_start_end_with.rb [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+          if receiver && args2.all?(&:pure?)
+            add_offense(
+              node,
+              :expression,
+              format(
+                MSG,
+                receiver: receiver.source,
```
Comment (bbatsov) [inline lib/rubocop/cop/performance/double_start_end_with.rb]: The `.source` is redundant.

Comment (bbatsov) [inline lib/rubocop/cop/performance/double_start_end_with.rb]: Forget what I said.

File: lib/rubocop/cop/performance/double_start_end_with.rb (hunk 4 above)
Comment (DNNX) [inline lib/rubocop/cop/performance/double_start_end_with.rb]: Fixed

File: spec/rubocop/cop/performance/double_start_end_with_spec.rb [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+  subject(:cop) { described_class.new }
+
+  before do
+    inspect_source(cop, source)
+  end
+
+  context 'two #start_with? calls' do
+    context 'with the same receiver' do
+      context 'all parameters of the second call are pure' do
+        let(:source) { 'x.start_with?(a, b) || x.start_with?("c", D)' }
```
Comment (bbatsov) [inline spec/rubocop/cop/performance/double_start_end_with_spec.rb]: As someone who does a bit of FP on the side, I think the usage of the term "pure" here is confusing. Basically you don' t want a function call as an argument, right?

File: lib/rubocop/cop/performance/double_start_end_with.rb [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+      #   var1 = ...
+      #   var2 = ...
+      #   str.end_with?(var1, var2)
+      class DoubleStartEndWith < Cop
+        MSG = 'Use `%{receiver}.%{method}(%{combined_args})` ' \
+              'instead of `%{original_code}`.'
+
+        def on_or(node)
+          receiver, method, args1, args2 = two_start_end_with_calls(node)
+          if receiver && args2.all?(&:pure?)
```
Comment (bbatsov) [inline lib/rubocop/cop/performance/double_start_end_with.rb]: I have to see what this function does, but I'm pretty sure it can't check whether a function is pure. :-)

File: spec/rubocop/cop/performance/double_start_end_with_spec.rb (hunk 6 above)
Comment (alexdowad) [inline spec/rubocop/cop/performance/double_start_end_with_spec.rb]: No -- "pure" means "pure", as in "has no side effects when evaluated". See the definition in `Node`.

Comment (DNNX) [inline spec/rubocop/cop/performance/double_start_end_with_spec.rb]: Yes that's right.

Comment (DNNX) [inline spec/rubocop/cop/performance/double_start_end_with_spec.rb]: Well, strictly speaking we can't allow local variables and constants too. There might be some pathalogical cases when `.to_str` have side effects:

```ruby
class Evil
  def initialize
    @state = 0
  end

  def to_str
    @state = (@state + 1) % 4
    @state.to_s
  end
end

EVIL = Evil.new

p "1".start_with?(EVIL, EVIL) || "1".start_with?(EVIL)
# => true
p "1".start_with?(EVIL, EVIL, EVIL)
# => false
```

Comment (bbatsov) [inline spec/rubocop/cop/performance/double_start_end_with_spec.rb]: `pure` doesn't mean "has not side effect when evaluated". It means that the same input always maps to the same output, that the only input is parameters (no global, instance vars, etc) and that they are no side effects. I'll check the definition.

Comment (alexdowad) [inline spec/rubocop/cop/performance/double_start_end_with_spec.rb]: If you can suggest a better name for this helper, please do so.

Comment (bbatsov) [inline spec/rubocop/cop/performance/double_start_end_with_spec.rb]: Maybe just `value?`.

Comment (DNNX) [inline spec/rubocop/cop/performance/double_start_end_with_spec.rb]: Sounds good to me. @alexdowad what do you think?

Comment (alexdowad) [inline spec/rubocop/cop/performance/double_start_end_with_spec.rb]: Hmm. Seems equally confusing with `pure?`, to be honest.

Comment (bbatsov) [inline spec/rubocop/cop/performance/double_start_end_with_spec.rb]: All the nodes you're checking seem to me like value nodes (and the complex nodes also reduce down to values). In general there are no method invocation checks and I really think that many people associate the term pure with functions and methods. I'm open to other naming suggestions, but I certainly think we can do better than the current name. I was recently reminded that I didn't like this name when I originally saw it, but the commit was a small part of a huge PR and I was too tired to go into relatively small details. Anyways, guess this is not related to the current PR.

**Comment by bbatsov** [comment]
Rebase this and I'll have it merged.

File: spec/rubocop/cop/performance/double_start_end_with_spec.rb (hunk 6 above)
Comment (alexdowad) [inline spec/rubocop/cop/performance/double_start_end_with_spec.rb]: Is `#no_side_effects?` too wordy?

Comment (bbatsov) [inline spec/rubocop/cop/performance/double_start_end_with_spec.rb]: My main issue is that such names imply we're dealing with something executable. I'd expect this to apply to send nodes, not to nodes in general. Naming is so hard...

**Comment by DNNX** [comment]
Rebased it. CI green, no conflicts.
